lib/bdev/nvme/blockdev_nvme.c

```diff
@@ -502,16 +503,59 @@ probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
 }
 
 static void
-blockdev_nvme_timeout_cb(void *cb_arg, struct spdk_nvme_ctrlr *ctrlr,
-			 struct spdk_nvme_qpair *qpair, uint16_t cid)
+abort_cb(void *arg, const struct spdk_nvme_cpl *cpl)
+{
+	/*
+	 * abort completion callback function. Currently, there is nothing to do
+	 * struct spdk_nvme_ctrlr *ctrlr = arg;
+	 */
+}
+
+static void
+blockdev_nvme_timeout_cb(struct spdk_nvme_ctrlr *ctrlr,
+			 struct spdk_nvme_qpair *qpair,
+			 void *cb_arg,
```

Comment:
So the case the timeout_count is addressing is handling an abort failure. There are, of course, several ways the abort can fail and at least two of those errors may not indicate a serious problem (the two you listed). Below is a bit of analysis on what I think the "right" thing to do is in response to an abort failure.

The original I/O, which I'll refer to as O, timed out. Then, the software sent an abort, A. A then completes with an error.

If A completes with the status "command not aborted", there does not seem to be any reason to believe that sending an abort again will work correctly. O has also already timed out, so waiting more time is not valuable. Therefore, the only path forward seems to be to do a controller reset or delete the queue pair.

If A completes with the status "abort command limit exceeded", then there is reason to believe that waiting some period of time and then trying to abort again may work. I think we can handle this case internally in the NVMe driver by adding an abort command queue to the admin queue pair. The limit on simultaneous outstanding aborts is known at start up time, so we can manage the queue in software and effectively avoid this error.

If A completes with any other error, the controller should be reset or the queue pair should be deleted.